zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -545,8 +553,16 @@ public void loadData() throws IOException, InterruptedException {
                         .filter(session -> zkDb.getSessionWithTimeOuts().get(session) == null)
                         .forEach(session -> killSession(session, zkDb.getDataTreeLastProcessedZxid()));
 
-        // Make a clean snapshot
-        takeSnapshot();
+        // Make a clean snapshot if needed
+        /* A snapshot is not needed when loading data during leader election. A new snapshot is not needed to return to
+         * the current state because the current state was reached by loading the data tree using an old snapshot.
+         * During leader election, there are no ongoing transactions that could be lost either. If a snapshot is needed
+         * to send to a learner during leader election, that snapshot will be taken as part of leader election, so
+         * snapshotting does not need to be done here as well.
+         */
```

Comment:
I believe this should answer your [question](https://github.com/apache/zookeeper/pull/2085#issuecomment-1994502832), right @horizonzy ?

Reply:
https://github.com/apache/zookeeper/blob/ee994fbca51f826e4b26d6a105866975d0007f6e/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java#L577

zookeeper's snapshot not only persists the dataTree but also persists the session and session expiration time, I'm concerned that if we don't take a snapshot here, the session data may not be accurate during the next restart.

@eolivelli /cc

Reply:
Thanks a lot for the info @horizonzy , is this session info used for something upon restart that is necessary for correctness? Is it mainly important for telemetry? If it is not needed for correctness could skipping this snapshotting step during leader election be made a configurable option?